packages/plugin-ext/src/plugin/preferences/configuration.ts

```diff
@@ -0,0 +1,575 @@
+/********************************************************************************
+ * Copyright (C) 2019 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+// copied from https://github.com/Microsoft/vscode/blob/6959ac313dbbe4a6b360de785c8ef4b548ebf8d7/src/vs/platform/configuration/common/configuration.ts,
+// https://github.com/Microsoft/vscode/blob/6959ac313dbbe4a6b360de785c8ef4b548ebf8d7/src/vs/platform/configuration/common/configurationModels.ts,
+// https://github.com/Microsoft/vscode/blob/6959ac313dbbe4a6b360de785c8ef4b548ebf8d7/src/vs/base/common/arrays.ts
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import URI from 'vscode-uri';
+import { WorkspaceExtImpl } from '../workspace';
+import { isObject } from '../../common/types';
+import { deepFreeze } from '@theia/core/lib/common/objects';
+import cloneDeep = require('lodash.clonedeep');
+
+/* tslint:disable:no-any */
+
```

Comment:
Unfortunately, that code from Monaco cannot be reused in plugin's part of API. In order to avoid opening a CQ, I got rid of copied code and implemented it on my own.